[SPARK-14480][SQL] Remove meaningless StringIteratorReader for CSV data source.

HyukjinKwon opened the pull request:
## What changes were proposed in this pull request?

This PR removes meaningless `StringIteratorReader` for CSV data source.

In `CSVParser.scala`, there is an `Reader` wrapping `Iterator` but there are two problems by this.

Firstly, it was actually not faster than processing line by line with Iterator due to additional logics to wrap `Iterator` to `Reader`.
Secondly, this brought a bit of complexity because it needs additional logics to allow every line to be read bytes by bytes. So, it was pretty difficult to figure out issues about parsing, (eg. SPARK-14103).

A benchmark was performed manually and the results were below:
- Original codes with Reader wrapping Iterator

| End-to-end (ns) | Parse Time (ns) |
| --- | --- |
| 14116265034 | 2008277960 |
- New codes with Iterator

| End-to-end (ns) | Parse Time (ns) |
| --- | --- |
| 13451699644 | 1549050564 |

For the details for the environment, dataset and methods, please refer the JIRA ticket.
## How was this patch tested?

Existing tests should cover this.

HyukjinKwon committed:
Remove StringIteratorReader

SparkQA commented:
**[Test build #60930 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60930/consoleFull)** for PR 13808 at commit [`978a728`](https://github.com/apache/spark/commit/978a7286255fb16f3ad4270d59cb58fe18c967c9).

SparkQA commented:
**[Test build #60930 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60930/consoleFull)** for PR 13808 at commit [`978a728`](https://github.com/apache/spark/commit/978a7286255fb16f3ad4270d59cb58fe18c967c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60930/
Test PASSed.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala:
```diff
@@ -28,11 +28,10 @@ import org.apache.spark.internal.Logging
+private[sql] class CsvReader(params: CSVOptions) {
 
-  protected lazy val parser: CsvParser = {
+  private lazy val parser: CsvParser = {
```
why is this lazy?

HyukjinKwon replied:
Oh, it seems it doesn't have to be. I will remove this and take a close look again.

HyukjinKwon committed:
Remove meaningless lazy and unused arguments

SparkQA commented:
**[Test build #60989 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60989/consoleFull)** for PR 13808 at commit [`5914689`](https://github.com/apache/spark/commit/59146894f7c8da2d70af984f6aeba14fe9c43ad7).

SparkQA commented:
**[Test build #60989 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60989/consoleFull)** for PR 13808 at commit [`5914689`](https://github.com/apache/spark/commit/59146894f7c8da2d70af984f6aeba14fe9c43ad7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60989/
Test PASSed.

HyukjinKwon commented:
@rxin Could you take another look please?

rxin commented:
This looks good. Let me merge it.

davies commented:
@HyukjinKwon @rxin This patch have a regression: A column that have escaped newline can't be correctly parsed anymore. Should we revert this patch or figure a way to fix that?

HyukjinKwon commented:
@davies, removed `StringIteratorReader` concatenates the lines in each iterator into reader in each partition IIRC.

New line in the column was not supported correctly up to my understanding because rows can spawn across multiple blocks. This is a similar problem that we have not supported multiple JSON lines before up to my knowledge. 

Currently,  we have some open PRs for dealing with multiple lines support by using something like `wholeTextFile` or dealing with each file as a multiple line json, which I think we could solve in that way if any of it is merged.

I guess we introduced several regression or behaviour changes when we porting. Would this be acceptable rather than supporting multiple lines one in this way?

HyukjinKwon commented:
FWIW, I remember I had a hard time to figure out https://issues.apache.org/jira/browse/SPARK-14103 where the issue itself was about quote but it ended up reading whole partition as a value.
